File: src/main/java/offer/technical/test/services/UserService.java
package offer.technical.test.services;

import lombok.RequiredArgsConstructor;
import offer.technical.test.errors.AlreadyExistsException;
import offer.technical.test.mapper.UserMapper;
import offer.technical.test.model.UserResource;
import offer.technical.test.repositories.UserRepository;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class UserService {

  private final UserRepository repository;
  private final UserMapper mapper;

  /**
   * Retrieve a user by username
   *
   * @param name username
   * @return a {@link UserResource} founded
   */
  public UserResource getUser(String name) {
    return mapper.userEntityToUserResource(repository.findOneByName(name));
  }

  /**
   * Register a user
   *
   * @param user a {@link UserResource} instance
   * @return a {@link UserResource} registered instance
   * @throws AlreadyExistsException if user already exists in database
   */
  public UserResource create(UserResource user) throws AlreadyExistsException {
    return mapper.userEntityToUserResource(repository.create(mapper.userResourceToUserEntity(user)));
  }

}

File: src/main/kotlin/offer/technical/test/services/UserService.kt
package offer.technical.test.services

import offer.technical.test.errors.AlreadyExistsException
import offer.technical.test.mapper.UserMapper
import offer.technical.test.model.UserResource
import offer.technical.test.repositories.UserRepository
import org.springframework.stereotype.Service

@Service
class UserService(var repository: UserRepository, var mapper: UserMapper) {

    /**
     * Retrieve a user by username
     *
     * @param name
     * @return user
     */
    fun getUser(name: String?): UserResource? {
        return mapper.userEntityUserResource(repository.findOneByName(name))
    }

    /**
     * Register a user
     *
     * @param user
     * @return user registered
     */
    @kotlin.Throws(AlreadyExistsException::class)
    fun create(user: UserResource?): UserResource? {
        return mapper.userEntityUserResource(mapper.userResourceToUserEntity(user)?.let { repository.create(it) })
    }

}